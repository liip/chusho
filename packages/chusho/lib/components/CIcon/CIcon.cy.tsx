import { ChushoUserOptions } from '../../types';

import { CIcon } from '.';

describe('CIcon', () => {
  it('applies local and global classes', () => {
    cy.mount(<CIcon id="palette" class="icon" data-test="icon" />, {
      global: {
        provide: {
          $chusho: {
            options: {
              components: {
                icon: {
                  class: 'config-icon',
                },
              },
            } as ChushoUserOptions,
          },
        },
      },
    });

    cy.get('[data-test="icon"]').should('have.class', 'icon config-icon');

    cy.getWrapper().then((wrapper) => {
      wrapper.setProps({ bare: true });

      cy.get('[data-test="icon"]').should('not.have.class', 'config-icon');
    });
  });

  it('applies the right attributes', () => {
    cy.mount(<CIcon id="palette" data-test="icon" />);

    cy.get('[data-test="icon"]')
      .should('have.attr', 'focusable', 'false')
      .and('have.attr', 'aria-hidden', 'true');
  });

  it('makes it visible and label it when `alt` is provided', () => {
    cy.mount(<CIcon id="palette" alt="Pick color" data-test="icon" />);

    cy.get('[data-test="icon"]')
      .should('have.attr', 'focusable', 'false')
      .should('not.have.attr', 'aria-hidden', 'true')
      .and('have.attr', 'aria-labelledby')
      .then((id) => {
        cy.get(`#${id}`).should('have.text', 'Pick color');
      });
  });

  it('combine sprite URL and ID as the icon source', () => {
    cy.mount(<CIcon id="palette" data-test="icon" />, {
      global: {
        provide: {
          $chusho: {
            options: {
              components: {
                icon: {
                  spriteUrl: '/icons.svg',
                },
              },
            } as ChushoUserOptions,
          },
        },
      },
    });

    cy.get('[data-test="icon"] use').should(
      'have.attr',
      'xlink:href',
      '/icons.svg#palette'
    );
  });

  it('supports icon ID without sprite URL', () => {
    cy.mount(<CIcon id="palette" data-test="icon" />, {
      global: {
        provide: {
          $chusho: {
            options: {
              components: {
                icon: {
                  spriteUrl: undefined,
                },
              },
            } as ChushoUserOptions,
          },
        },
      },
    });

    cy.get('[data-test="icon"] use').should(
      'have.attr',
      'xlink:href',
      '#palette'
    );
  });

  it('sizes the icon based on the config width/height and the scale prop', () => {
    cy.mount(<CIcon id="palette" scale={1} data-test="icon" />, {
      global: {
        provide: {
          $chusho: {
            options: {
              components: {
                icon: {
                  width: 16,
                  height: 16,
                },
              },
            } as ChushoUserOptions,
          },
        },
      },
    });

    // Scale defaults to 1
    cy.get('[data-test="icon"]').should('have.attr', 'width', '16');
    cy.get('[data-test="icon"]').should('have.attr', 'height', '16');

    cy.getWrapper().then((wrapper) => {
      wrapper.setProps({ scale: 2 });

      cy.get('[data-test="icon"]').should('have.attr', 'width', '32');
      cy.get('[data-test="icon"]').should('have.attr', 'height', '32');
    });
  });
});
