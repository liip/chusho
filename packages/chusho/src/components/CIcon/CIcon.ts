import Vue, { VNode, VNodeData } from 'vue';
import { mergeData } from 'vue-functional-data-merge';
import uuid from '../../utils/uuid';

interface IconProps {
  id: string;
  scale: number;
  alt: string;
}

export default Vue.extend<IconProps>({
  name: 'CIcon',

  functional: true,

  props: {
    /**
     * The id of the symbol (icon) to display from within the sprite.
     */
    id: {
      type: String,
      required: true,
    },
    /**
     * The scale relative to the dimensions defined in the config (icon.width, icon.height).
     */
    scale: {
      type: Number,
      default: 1,
    },
    /**
     * Provides an alternate text to describe the icon meaning in the context its used.
     * In cases where an icon would be used without any label close by, this is important
     * to provide for accessibility.
     *
     * Example: imagine a lonely trash icon within a button dedicated to delete an article,
     * this prop should be set to a value like "Delete article".
     */
    alt: {
      type: String,
      default: '',
    },
  },

  render(h, { props, data, parent }): VNode {
    const iconConfig = parent?.$chusho?.options?.components?.icon;
    const componentData: VNodeData = {
      attrs: {
        focusable: 'false',
      },
      class: iconConfig?.class,
      style: {
        width: `${(iconConfig?.width || 24) * props.scale}px`,
        height: `${(iconConfig?.height || 24) * props.scale}px`,
      },
    };
    const uid = `chusho-icon-${uuid()}`;

    if (componentData.attrs) {
      if (props.alt) {
        componentData.attrs['aria-labelledby'] = uid;
      } else {
        componentData.attrs['aria-hidden'] = true;
      }
    }

    return h('svg', mergeData(data, componentData), [
      props.alt &&
        h(
          'title',
          {
            attrs: {
              id: uid,
            },
          },
          [props.alt]
        ),
      h('use', {
        key: props.id,
        attrs: {
          'xlink:href': `${iconConfig?.spriteUrl}#${props.id}`,
        },
      }),
    ]);
  },
});
