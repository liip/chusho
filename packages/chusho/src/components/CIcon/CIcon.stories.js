import CIcon from './CIcon';

export default {
  title: 'Components|Icon',
  component: CIcon,
  parameters: {
    componentSubtitle:
      'Display a scalable SVG icon from a pre-existing sprite.',
    options: {
      componentConfig: [
        {
          name: 'spriteUrl',
          type: { summary: 'string' },
          description:
            'Load the sprite from a remote URL. This technique doesnâ€™t work on Internet Explorer, but it can be polyfilled with [svgxuse](https://github.com/Keyamoon/svgxuse). Learn more about [SVG Sprites](https://css-tricks.com/svg-sprites-use-better-icon-fonts/) on CSS-Tricks.',
        },
        {
          name: 'width',
          type: { summary: 'number' },
          description: 'Icon viewBox width.',
          defaultValue: { summary: 24 },
          required: true,
        },
        {
          name: 'height',
          type: { summary: 'number' },
          description: 'Icon viewBox height.',
          defaultValue: { summary: 24 },
          required: true,
        },
        {
          name: 'class',
          type: { summary: 'string' },
          description: 'Class applied to all Icon instances.',
        },
      ],
    },
  },
};

export const Default = () => ({
  components: { CIcon },
  template: `<CIcon id="palette" />`,
});

export const CustomScale = () => ({
  components: { CIcon },
  template: `<CIcon id="picture" :scale="0.5" />`,
});

export const WithAlternateText = () => ({
  components: { CIcon },
  template: `<CIcon id="picture" alt="Zoom picture" />`,
});
