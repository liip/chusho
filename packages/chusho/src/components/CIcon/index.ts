import CIcon from './CIcon.vue';

export { CIcon };
export default CIcon;
