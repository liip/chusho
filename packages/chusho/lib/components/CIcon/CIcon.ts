import { defineComponent, h, mergeProps } from 'vue';

import componentMixin from '../mixins/componentMixin';

import useComponentConfig from '../../composables/useComponentConfig';

import { generateConfigClass } from '../../utils/components';
import uuid from '../../utils/uuid';

export default defineComponent({
  name: 'CIcon',

  mixins: [componentMixin],

  inheritAttrs: false,

  props: {
    /**
     * The id of the symbol (icon) to display from the sprite.
     */
    id: {
      type: String,
      required: true,
    },

    /**
     * Multiply the width/height defined in the config to change the icon display size.
     */
    scale: {
      type: Number,
      default: 1,
    },

    /**
     * Provides an alternate text to describe the icon meaning in the context its used.
     * In cases where an icon would be used without any label close by, this is important
     * to provide for accessibility.
     *
     * Example: imagine a lonely trash icon within a button dedicated to delete an article,
     * this prop should be set to a value like "Delete article".
     */
    alt: {
      type: String,
      default: '',
    },
  },

  setup() {
    return {
      config: useComponentConfig('icon'),
    };
  },

  render() {
    const elementProps: Record<string, unknown> = mergeProps(this.$attrs, {
      focusable: 'false',
      width: `${(this.config?.width || 24) * this.scale}`,
      height: `${(this.config?.height || 24) * this.scale}`,
      ...generateConfigClass(this.config?.class, this.$props),
    });
    const id = uuid('chusho-icon');

    if (this.alt) {
      elementProps['aria-labelledby'] = id;
    } else {
      elementProps['aria-hidden'] = true;
    }

    return h('svg', elementProps, [
      this.alt &&
        h(
          'title',
          {
            id,
          },
          [this.alt]
        ),
      h('use', {
        key: this.id,
        'xlink:href': `${this.config?.spriteUrl || ''}#${this.id}`,
      }),
    ]);
  },
});
