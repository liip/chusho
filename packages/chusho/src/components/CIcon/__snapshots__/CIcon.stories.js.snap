// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Components|Icon Custom Scale 1`] = `
<svg
  aria-hidden="true"
  class="inline-block align-middle pointer-events-none fill-current"
  focusable="false"
  style="width: 24px; height: 24px;"
>
  <use
    xlink:href="icons.svg#picture"
  />
</svg>
`;

exports[`Storyshots Components|Icon Default 1`] = `
<svg
  aria-hidden="true"
  class="inline-block align-middle pointer-events-none fill-current"
  focusable="false"
  style="width: 48px; height: 48px;"
>
  <use
    xlink:href="icons.svg#palette"
  />
</svg>
`;

exports[`Storyshots Components|Icon With Alternate Text 1`] = `
<svg
  aria-labelledby="chusho-icon-2"
  class="inline-block align-middle pointer-events-none fill-current"
  focusable="false"
  style="width: 48px; height: 48px;"
>
  <title
    id="chusho-icon-2"
  >
    Zoom picture
  </title>
  <use
    xlink:href="icons.svg#picture"
  />
</svg>
`;
